app.py

```diff
@@ -1,150 +1,269 @@
+        # streamlit re-run and comparing the seemingly same class instances does not work then
+        st.session_state['model'] = model_current.value
+        st.session_state['toolkit'] = toolkit_current.value
+    else:
+        agent = st.session_state.agent
+
+    assert isinstance(agent, AgentExecutor)
+
+    chat_history = StreamlitChatMessageHistory(key=_ST_CHAT_HiSTORY_KEY)
+    st.sidebar.button('Clear Chat', on_click=lambda: chat_history.clear())
+
+    # this makes columns only as wide as are the buttons, see https://stackoverflow.com/a/77332142
+    st.markdown("""
+            <style>
+                div[data-testid="column"] {
+                    width: fit-content !important;
+                    flex: unset;
+                }
+                div[data-testid="column"] * {
+                    width: fit-content !important;
+                }
+            </style>
+            """, unsafe_allow_html=True)
+
+    chat = st.container()
+    with chat:
+        if len(chat_history.messages) == 0:
+            chat_history.add_ai_message(ai_intro)
+
+        for msg in chat_history.messages:
+            st.chat_message(msg.type).write(msg.content)
+
+    if user_input := st.chat_input():
+        with chat:
+            st.session_state.pop(_ST_TRANS_QUERY, None)
+            st.chat_message('user').write(user_input)
+            response = _call_agent(agent, user_input)
+            st.chat_message('ai').write(response['output'])
+
+    # get the output of the last message from the agent
+    if len(chat_history.messages) > 1:
+        last_output_message = chat_history.messages[-1].content
+        if re.findall(r'```sql\n(.*?)\n```', last_output_message, re.DOTALL):
+            with chat:
+                col1, col2 = st.columns(2)
+
+            def _execute_sql_handler():
+                with chat:
+                    _call_agent(agent, 'Run the last query!')
+
+            def _create_transformation_handler():
+                st.session_state[_ST_TRANS_QUERY] = query
+
+            query = re.findall(r'```sql\n(.*?)\n```', last_output_message, re.DOTALL)[0]
+            col1.button(
+                'Execute SQL',
+                on_click=_execute_sql_handler,
+                use_container_width=True,
+                disabled=bool(st.session_state.get(_ST_TRANS_QUERY)))
+            col2.button(
+                'Create Transformation',
+                on_click=_create_transformation_handler,
+                use_container_width=True,
+                disabled=bool(st.session_state.get(_ST_TRANS_QUERY)))
+
+    if st.session_state.get(_ST_TRANS_QUERY):
+        query = st.session_state.get(_ST_TRANS_QUERY)
+        with chat:
+            with st.form(key='transformation_details', clear_on_submit=False):
+                st.session_state['ai_config_details'] = generate_config_details(chat_history.messages[-2:])
```

Comment:
I could be wrong, but it seems unnecessary to store the config details to the streamlit session. You immediately consume those details no the lines below and nowhere else. Perhaps an ordinary local variable could do the same job, or not?

Reply:
I had to rework this a little bit to make the form show all its internals (text boxes, buttons) even while the LLM is preparing the transformation details and also to show the spinner which explains that the form is being prepared. See 11f5fc0.